### ISSUE TITLE
java.lang.String cannot be cast to scala.collection.TraversableLike

### ISSUE THREAD

**Issue by arjunsingla** [JohnSnowLabs/spark-nlp#607]
## Code

<img width="1083" alt="Screen Shot 2019-08-23 at 12 14 37 PM" src="https://user-images.githubusercontent.com/6709889/63618384-cf028980-c5a0-11e9-9884-096167e34d19.png">
<img width="1028" alt="Screen Shot 2019-08-23 at 12 14 48 PM" src="https://user-images.githubusercontent.com/6709889/63618392-d0cc4d00-c5a0-11e9-9932-42fca4ad261f.png">
<img width="1081" alt="Screen Shot 2019-08-23 at 12 14 57 PM" src="https://user-images.githubusercontent.com/6709889/63618405-d32ea700-c5a0-11e9-8780-b11ecf75d78f.png">
<img width="1086" alt="Screen Shot 2019-08-23 at 12 15 03 PM" src="https://user-images.githubusercontent.com/6709889/63618406-d5910100-c5a0-11e9-854a-0607531f6e35.png">
<img width="1011" alt="Screen Shot 2019-08-23 at 12 15 11 PM" src="https://user-images.githubusercontent.com/6709889/63618411-d9248800-c5a0-11e9-9642-17adafb4b8d4.png">
<img width="1026" alt="Screen Shot 2019-08-23 at 12 15 18 PM" src="https://user-images.githubusercontent.com/6709889/63618418-db86e200-c5a0-11e9-8563-edef01ad9412.png">
<img width="962" alt="Screen Shot 2019-08-23 at 12 15 23 PM" src="https://user-images.githubusercontent.com/6709889/63618426-df1a6900-c5a0-11e9-9574-59ab8b4af951.png">

**Comment by maziyarpanahi** [comment]
Is that one of our notebook? We fix this bug in 2.2.0 however you should use either “document” or “sentence” as inputCols for all the annotators. You can’t use document for Tokenizer and sentence for the reset. It needs the position of the tokens whether in the document or in each sentence. Change Tokenizer and nerConverter to use sentence and you’ll be OK.

**Comment by arjunsingla** [comment]
Thank you for your quick response. I changed the Tokenizer and nerConverter to use 'sentence' instead of document for the input columns, so now my code looks like:

```
documentAssembler = DocumentAssembler()\
  .setInputCol("TextField")\
  .setOutputCol("document")

sentenceDetector = SentenceDetector()\
    .setInputCols(['document'])\
    .setOutputCol('sentence')

tokenizer = Tokenizer()\
  .setInputCols(["sentence"])\
  .setOutputCol("token")

nerTagger = NerDLApproach()\
  .setInputCols(["sentence", "token"])\
  .setLabelColumn("Fix")\
  .setOutputCol("ner")\
  .setMaxEpochs(109)\
  .setRandomSeed(0)\
  .setVerbose(2)

converter = NerConverter()\
  .setInputCols(["sentence", "token", "ner"])\
  .setOutputCol("ner_span")
    
finisher = Finisher() \
  .setInputCols(["ner_span"])\
    .setIncludeMetadata(True)

pipeline = Pipeline(
  stages = [
    documentAssembler,
    sentenceDetector,
    tokenizer,
    nerTagger,
    converter,
    finisher
  ]
)
```

I am still getting the same error. Any ideas?

**Comment by arjunsingla** [comment]
Update: I even updated to 2.2.0, and am getting the same error. Any ideas?

**Comment by maziyarpanahi** [comment]
Training NerDLApproach needs 3 inputs and one is WordEmbeddings which you are missing from the pipeline. Add WordEmbeddings right before NerDLApproach and use it as an input.
Also, I don’t see you using CoNLL() class for converting your training dataset.

Check out some examples for start: https://github.com/JohnSnowLabs/spark-nlp-workshop

**Comment by arjunsingla** [comment]
Looking at the github example, I am a little confused. I was hoping you could answer some questions.

My goal is to look at work order notes for fixing machines, and use nlp to extract the names of parts that were used in the fix from them. My current dataset, which i have loaded into a spark dataframe, contains the work order notes in one column (a paragraph or so) and a list of parts that are mentioned in the notes in the second column. Some of the parts are multiple words (ie, X ray tube). 

I was under the impression that I could just train the model on this dataframe. Where do the Word Embeddings and CoNLL() class come in? Do I need to format my data differently? The example shows that you used glove word embeddings. Can I use this as is?

Thanks.

**Comment by maziyarpanahi** [comment]
Sure, before I answer your question I need to know if what you are trying to do is possible. Are the parts associated with the whole work order note or is it word by word? For instance, is X Ray Tube refers to the whole entry or points to a specific token in the work orders?

**Comment by arjunsingla** [comment]
The parts are associated with the whole work order. They are embedding inside of it. For example, the work order might say "John fixed the X ray tube of the machine at 5:48 in the afternoon. Judy oversaw him", and then what I would want to extract from that text would the be "X Ray tube" part.

Thanks!

**Comment by maziyarpanahi** [comment]
As I guessed. That’s not NER (Named Entity Recognition). As the name suggests, each token in a sentence must have a label. What you need is a multi-label classifier by using Spark ML functions. Perhaps OneVsRestClassifier with NaiveBayesClassifier where you make vectors out of your text and other important parameters and train them with your labels. 
You can use Spark NLP transformers to extract good text but for training the last model you need classification algorithms which can be found in Spark ML. (You don’t have to use Spark NLP at all for this but in case you want to there are examples in Strata directory of the repo I gave you that shows how to integrate with Spark ML)

**Comment by arjunsingla** [comment]
I see.

That seems very complex. Would it be possible if I could use Spark nlp to extract each word in a part? For example, if the work order said. "John fixed the X ray tube of the machine at 5:48 in the afternoon. Judy oversaw him", and I wanted to extract "X", "ray", and "tube", would this be possible?  
I honestly am just starting with NLP and I appreciate your help

**Comment by maziyarpanahi** [comment]
Are the parts names always being mention inside the order?

**Comment by arjunsingla** [comment]
Yes, they are.

**Comment by maziyarpanahi** [comment]
Then you don’t need to train anything! Training a model means you need to predict the output based on unseen input. Everything here is static and can be extracted by rules like regex. 
You should remove everything after Tokenizer and just add TextMatcher annotator. There you just point to a text file which each line is a name of the parts you wish to be extracted, then it extracts those from the order column and add another column which you can call it parts. Very simple with few lines of code :)

**Comment by arjunsingla** [comment]
Give me a second, let me think about this :)

**Comment by arjunsingla** [comment]
I believe I misspoke. The part names are always being mentioned inside the order on the current dataset. But the dataset does not contain all the possible part names for the machine. I want my pipeline to be able to be able to extract not just the part names from this dataset, but to realize that some words correspond to parts that it has not seen before, so it can be used on future work orders. I hope that clears it up. Do you have any thoughts?

**Comment by arjunsingla** [comment]
For example, after training on this dataset, if a work order says, "Fixed the generator," my model should be able to figure out that generator is a part - even though the generator part was not mentioned on my current dataset.

**Comment by maziyarpanahi** [comment]
The two solutions I offered you could get you close to what you need. However, even the parts being dynamic and could be anything is something super complicated. I don’t think you can just have few lines and small data set to achieve that. You can have a look at Part of Speech or dependency parsers and see if you can get the parts based on its position or relations in the sentence.
But in terms of training a model to guess a class/cat that it’s never seen, I don’t think it’s possible with Spark nor Spark NLP, sorry.

**Comment by maziyarpanahi** [comment]
Yes, he is the lead engineer and responsible for this library. The dataset for training is what makes something possible or impossible. In theory it is possible of you have a very good and scientifically made dataset for NER and instead of diseases it could be part names. Exactly this photo:
![image](https://user-images.githubusercontent.com/5762953/63720414-67259a80-c84f-11e9-8684-d0762b938ad6.png)

However, you don’t have a data set that is useful for training these types of NLP tasks. In fact your dataset is not sequential labeling it is pure classification where the entire entry has one or more tags (part names). This falls under data preparation and ML engineering where you be able to train a model that does what you need. Which requires a data scientist working with you and your business to understand and use every parameter for more accurate model.

Without that, you should explore other options as I mentioned which may not look sophisticated or flexible but it’s something. Look into grammatical relations or Part of Speech if the sentences are usually the same structure. What you want to achieve is absolutely impossible for NER with your dataset.

**Comment by arjunsingla** [comment]
Hi!

Sorry for the late reply. I took the last week to convert my work order dataset into a conll one, as you mentioned. I then labled individual words that corresponded to part names (ie "x", "ray", "tube) instead of trying to extract the whole thing (ie "x ray tube"). After building this dataset, I trained the John Snow labs named entity recognition model on this dataset, and I am happy to report that it works fairly well (with no errors!). Thank you so much for your help.
